### PR TITLE
feat: add custom and hp tuning

### PR DESCRIPTION
**Pull request by sasha-gitg**

### CODE REVIEW

File: google/cloud/aiplatform/hyperparameter_tuning.py [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+    ):
+
+        self.conditional_parameter_spec = conditional_parameter_spec
+        self.parent_values = parent_values
+
+    @property
+    @classmethod
+    @abc.abstractmethod
+    def _proto_parameter_value_class(self) -> proto.Message:
+        """The proto represenation of this parameter."""
```
Comment (ivanmkc) [inline google/cloud/aiplatform/hyperparameter_tuning.py:49]: Spelling: Representation

File: google/cloud/aiplatform/jobs.py [hunk 2]
```diff
@@ -788,10 +879,561 @@ class DataLabelingJob(_Job):
...
+}
+
+_MEASUREMENT_SELECTION_TO_PROTO_VALUE = {
+    "best": gca_study_compat.StudySpec.MeasurementSelectionType.BEST_MEASUREMENT,
+    "last": gca_study_compat.StudySpec.MeasurementSelectionType.LAST_MEASUREMENT,
+}
+
+
+class HyperparameterTuningJob(_RunnableJob):
+    """AI Pltatform(Unified) HyperparameterTuning Job"""
```
Comment (ivanmkc) [inline google/cloud/aiplatform/jobs.py:1186]: Spelling and spacing issues.

File: google/cloud/aiplatform/jobs.py [hunk 3]
```diff
@@ -788,10 +879,561 @@ class DataLabelingJob(_Job):
...
+            service_account (str):
+                Optional. Specifies the service account for workload run-as account.
+                Users submitting jobs must have act-as permission on this run-as account.
+            network (str):
+                Optional. The full name of the Compute Engine network to which the job
+                should be peered. For example, projects/12345/global/networks/myVPC.
+                Private services access must already be configured for the network.
+                If left unspecified, the job is not peered with any network.
+            timeout (int):
+                The maximum job running time in seconds. The default is 7 days.
```
Comment (ivanmkc) [inline google/cloud/aiplatform/jobs.py:1399]: Missing note that it is optional.

File: google/cloud/aiplatform/jobs.py [hunk 4]
```diff
@@ -788,10 +879,561 @@ class DataLabelingJob(_Job):
...
+        """
+
+        if service_account:
+            self._gca_resource.trial_job_spec.service_account = service_account
+
+        if network:
+            self._gca_resource.trial_job_spec.network = network
+
+        if timeout or restart_job_on_worker_restart:
+            timeout = duration_pb2.Duration(seconds=timeout) if timeout else None
```
Comment (ivanmkc) [inline google/cloud/aiplatform/jobs.py:1417]: timeout was type `Int` but then is changed to a `duration_pb2.Duration` type. Is that a good practice?

File: google/cloud/aiplatform/jobs.py [hunk 5]
```diff
@@ -788,10 +879,561 @@ class DataLabelingJob(_Job):
...
+            for parameter_id, parameter in parameter_spec.items()
+        ]
+
+        study_spec = gca_study_compat.StudySpec(
+            metrics=metrics,
+            parameters=parameters,
+            algorithm=_SEARCH_ALGORITHM_TO_PROTO_VALUE[search_algorithm],
+            measurement_selection_type=_MEASUREMENT_SELECTION_TO_PROTO_VALUE[
+                measurement_selection
+            ],
```
[the remark this answers is not in the document]
Comment (vinnysenthil) [reply google/cloud/aiplatform/jobs.py:1363]: Ignore this if `None: ALGORITHM_UNSPECIFIED` is added to `_SEARCH_ALGORITHM_TO_PROTO_VALUE`

**Review by vinnysenthil — APPROVED** [review]
Looks great, thanks for getting this in time for the release 👏🏽  
Just a few nits and a requested change around `search_algorithms`.

- [x] Code review
- [ ] Successful manual test of HPT
- [x] Successful manual test of CustomJob

File: google/cloud/aiplatform/base.py [hunk 6]
```diff
@@ -543,6 +566,11 @@ def update_time(self) -> datetime.datetime:
         self._sync_gca_resource()
         return self._gca_resource.update_time
 
+    @property
+    def gca_resource(self) -> proto.Message:
+        """The underlying resource proto represenation."""
+        return self._gca_resource
```
[the remark this answers is not in the document]
Comment (sasha-gitg) [reply google/cloud/aiplatform/base.py:572]: Customers may need a field we haven't exposed. This allows us to provide code that doesn't require referencing a private attribute.

**Comment by sasha-gitg** [comment]
Thanks for the awesome review Vinny!

**Pull request merged**